Implements GatewayProven handler

pgev opened the pull request:
Fixes #297

Commit:
Drafts GatewayProvenHanlder

Commit:
Improves documentation

pgev commented:
The implementation is ready (draft PR is linked to the ticket). Tests are missing.
Moving back as someone else can pick.

0xsarvesh commented on src/m1_facilitator/handlers/GatewayProvenHandler.ts:
```diff
@@ -0,0 +1,112 @@
+   *        if `record.blockNumber` is greater than `remoteGatewayLastProvenBlockNumber`
+   *        of the stored value in the repository.
+   */
+  public async handle(records: any[]): Promise<void> {
```
We can remove any with specific type.

pgev replied:
Do we have a typescript type for this or you are proposing to create an interface for that here?

0xsarvesh replied:
Proposing a new interface. 

https://github.com/mosaicdao/facilitator/blob/1c9413af912815c908d4e583042fdf2842460bd2/src/m1_facilitator/handlers/AvailableStateRootsHandler.ts#L21

pgev replied:
Done.

0xsarvesh commented on src/m1_facilitator/handlers/GatewayProvenHandler.ts:
```diff
@@ -0,0 +1,112 @@
+      const gatewayModelRecord: Gateway | null = await this.gatewayRepository.get(
+        gatewayProvenEntity.gatewayAddress,
+      );
+      assert(
```
I am not sure we want to assert here: 
1. This condition will occur when events from other gateways will appear. This can happen when subgraph deployment has multiple gateway address or wrong address. This essentially is not issue with facilitator codebase. Graph and subgraphs are external systems to facilitator. 
2. If we consider overall architecture,  with this assert, it will throw an error and stop of this execution cycle(callback from subscription). It will not stop facilitator app.  Which also means this will stop execution of other handlers.
 
I think this should just log and silently ignore these records.

pgev replied:
I see you point. I've updated code by logging (as a warning) and leaving function on that point.

0xsarvesh commented on src/m1_facilitator/handlers/GatewayProvenHandler.ts:
```diff
@@ -0,0 +1,112 @@
+      const gatewayProvenEntity = new GatewayProvenEntity(
+        record.contractAddress,
+        record.remoteGateway,
+        record.blockNumber,
```
type of blocknumber from the subgraph will be string.

pgev replied:
I've updated implementation and changed GatewayProvenEntity constructor to accept string (it creates a BigNumber object from the string).

0xsarvesh commented on src/m1_facilitator/handlers/GatewayProvenHandler.ts:
```diff
@@ -0,0 +1,112 @@
+      await this.gatewayRepository.save(updatedGatewayModel);
+    });
+
+    await Promise.all(savePromises);
```
what will happen when two Gateway proven entity records appear in handle method? 
I think, it will call save twice which intern will trigger services twice. It should only save once only the higher block height.

pgev replied:
That's a great point. The current implementation will fail. I've added mutex to guard that.

0xsarvesh replied:
I was proposing to only update block height(remoteGatewayLastProvenBlockNumber)  and ignore lower block height.

pgev replied:
Addressed (in a different way).

0xsarvesh reviewed (commented):
Looks nice 👍  
Some questions and feedbacks.

Commit:
Merges with 'upstream/develop'

Commit:
Adds Mutex to handle multiple GatewatProven events

- Ignores (logs & returns) in case if there is no gateway
  model in GatewayRepository for a received GatewayProven
  entity.

Commit:
Explains mutex usage

gulshanvasnani commented on src/m1_facilitator/handlers/GatewayProvenHandler.ts:
```diff
@@ -0,0 +1,129 @@
+      // As potentially multiple GatewayProven events for the same Gateway
+      // can appear in handler, we lock 'check should we upsert and upsert itself'
+      // part.
+      const release = await this.mutex.acquire();
```
I think if going by `mutex` approach, then acquiring of `mutex` must be done before fetching the gateway record i.e. at line 83.

pgev replied:
Addressed (in a different way).

Commit:
Mutexes AnchorRepository and GatewayRepository save functions

Commit:
Merges w/ 'upstream/develop'

Commit:
Mutexes AnchorRepository and GatewayRepository save functions

Commit:
Adds GatewayProvenHandler::handle() unit tests

Commit:
Merge remote-tracking branch 'upstream/develop' into gh-297_gateway-proven-handler

deepesh-kn commented on src/m1_facilitator/handlers/GatewayProvenHandler.ts:
```diff
@@ -0,0 +1,111 @@
+   * handle() function accepts an array subgraph gateway-proven entities.
+   * If there is no model in the gateway repository matching to a gateway
+   * global address of an entity, the function logs a warning without failing.
+   * Otherwise, it updates a model in the gateway repository matchin to a
```
```suggestion
   * Otherwise, it updates a model in the gateway repository matching to a
```

pgev replied:
Done.

deepesh-kn commented on src/m1_facilitator/handlers/GatewayProvenHandler.ts:
```diff
@@ -0,0 +1,111 @@
+   */
+  public async handle(entities: GatewayProvenEntityInterface[]): Promise<void> {
+    const savePromises = entities.map(async (entity): Promise<void> => {
+      const gatewayProven = new GatewayProven(entity);
```
Why is the significance of the `GatewayProven` class?
We already have `GatewayProvenEntityInterface`

pgev replied:
GatewayProvenEntityInterface represents subgraph entity (like, blockNumber type is `string`). However, GatewayProven class does type checks and conversion of that entity.

Commit:
Fixes type

deepesh-kn commented:
We should add one more test related to `mutex`, to be sure that it is providing the expected result.

deepesh-kn commented:
> We should add one more test related to `mutex`, to be sure that it is providing the expected result.

Let us see how this behaves in the system test. Merging this.
